src/hlsBinaries.ts

```diff
@@ -206,44 +215,54 @@ export async function findHaskellLanguageServer(
         fs.mkdirSync(storagePath);
     }
 
-    const manageHLS = workspace.getConfiguration('haskell').get('manageHLS') as boolean;
+    if (manageHLS === null) { // plugin needs initialization
+      const promptMessage =
+          'How do you want the extension to manage/discover HLS?';
+
+      const decision = await window.showInformationMessage(promptMessage, 'system ghcup (recommended)', 'internal ghcup', 'PATH') || null;
+      if (decision === 'system ghcup (recommended)') {
+        manageHLS = 'system-ghcup';
+      } else if (decision === 'internal ghcup') {
+        manageHLS = 'internal-ghcup';
+      } else if (decision === 'PATH') {
+        manageHLS = 'PATH';
+      }
+      if (manageHLS !== null) {
+        workspace.getConfiguration('haskell').update('manageHLS', manageHLS, ConfigurationTarget.Global);
+      }
+    }
 
-    if (!manageHLS) {
-        const wrapper = findHLSinPATH(context, logger, folder);
-        if (!wrapper) {
-            const msg = 'Could not find a HLS binary! Consider installing HLS via ghcup or set "haskell.manageHLS" to true';
-            window.showErrorMessage(msg);
-            throw new Error(msg);
+    if (manageHLS === 'PATH' || manageHLS === null) {
+        if (workspace.getConfiguration('haskell').get('serverExecutablePath') as string !== '') {
+          return findServerExecutable(context, logger, folder);
         } else {
-            return wrapper;
+          return findHLSinPATH(context, logger, folder);
         }
     } else {
-        // permissively check if we have HLS installed
-        // this is just to avoid a popup
-        let wrapper = await callGHCup(context, logger,
-            ['whereis', 'hls'],
-            undefined,
-            false,
-            (err, stdout, _stderr, resolve, _reject) => { err ? resolve('') : resolve(stdout?.trim()); }
-        );
-        if (wrapper === '') {
-            // install recommended HLS... even if this doesn't support the project GHC, because
-            // we need a HLS to find the correct project GHC in the first place
-            await callGHCup(context, logger,
+        // we manage HLS, make sure ghcup is installed/available
+        await getGHCup(context, logger);
+
+        // get a preliminary hls wrapper for finding project GHC version,
+        // later we may install a different HLS that supports the given GHC
+        let wrapper = await getLatestHLSfromGHCup(context, storagePath, logger).then(e =>
+          (e === null)
+          ?  callGHCup(context, logger,
```

Comment:
Here's some considerable cleanup... now we pick a proper hls-wrapper executable (e.g. a compiled one) for the first pass (the one that just discovers project GHC), not just the `recommended`... then the next pass will possibly downgrade the wrapper to match the project GHC.